File: library/db/src/main/java/org/keynote/godtools/android/db/TranslationMapper.java
package org.keynote.godtools.android.db;

import android.content.ContentValues;
import android.database.Cursor;

import org.cru.godtools.model.Language;
import org.cru.godtools.model.Translation;

import androidx.annotation.NonNull;

import static org.cru.godtools.model.Translation.DEFAULT_LAST_ACCESSED;
import static org.cru.godtools.model.Translation.DEFAULT_PUBLISHED;
import static org.cru.godtools.model.Translation.DEFAULT_VERSION;
import static org.keynote.godtools.android.db.Contract.TranslationTable.COLUMN_DESCRIPTION;
import static org.keynote.godtools.android.db.Contract.TranslationTable.COLUMN_DOWNLOADED;
import static org.keynote.godtools.android.db.Contract.TranslationTable.COLUMN_LANGUAGE;
import static org.keynote.godtools.android.db.Contract.TranslationTable.COLUMN_LAST_ACCESSED;
import static org.keynote.godtools.android.db.Contract.TranslationTable.COLUMN_MANIFEST;
import static org.keynote.godtools.android.db.Contract.TranslationTable.COLUMN_NAME;
import static org.keynote.godtools.android.db.Contract.TranslationTable.COLUMN_PUBLISHED;
import static org.keynote.godtools.android.db.Contract.TranslationTable.COLUMN_TAGLINE;
import static org.keynote.godtools.android.db.Contract.TranslationTable.COLUMN_TOOL;
import static org.keynote.godtools.android.db.Contract.TranslationTable.COLUMN_VERSION;

final class TranslationMapper extends BaseMapper<Translation> {
    @Override
    protected void mapField(@NonNull final ContentValues values, @NonNull final String field,
                            @NonNull final Translation translation) {
        switch (field) {
            case COLUMN_TOOL:
                values.put(field, translation.getToolCode());
                break;
            case COLUMN_LANGUAGE:
                values.put(field, serialize(translation.getLanguageCode()));
                break;
            case COLUMN_VERSION:
                values.put(field, translation.getVersion());
                break;
            case COLUMN_NAME:
                values.put(field, translation.getName());
                break;
            case COLUMN_DESCRIPTION:
                values.put(field, translation.getDescription());
                break;
            case COLUMN_TAGLINE:
                values.put(field, translation.getTagline());
                break;
            case COLUMN_MANIFEST:
                values.put(field, translation.getManifestFileName());
                break;
            case COLUMN_PUBLISHED:
                values.put(field, translation.isPublished());
                break;
            case COLUMN_DOWNLOADED:
                values.put(field, translation.isDownloaded());
                break;
            case COLUMN_LAST_ACCESSED:
                values.put(field, serialize(translation.getLastAccessed()));
                break;
            default:
                super.mapField(values, field, translation);
                break;
        }
    }

    @NonNull
    @Override
    protected Translation newObject(@NonNull final Cursor c) {
        return new Translation();
    }

    @NonNull
    @Override
    public Translation toObject(@NonNull final Cursor c) {
        final Translation translation = super.toObject(c);

        translation.setToolCode(getString(c, COLUMN_TOOL));
        translation.setLanguageCode(getLocale(c, COLUMN_LANGUAGE, Language.INVALID_CODE));
        translation.setVersion(getInt(c, COLUMN_VERSION, DEFAULT_VERSION));
        translation.setName(getString(c, COLUMN_NAME, null));
        translation.setDescription(getString(c, COLUMN_DESCRIPTION, null));
        translation.setTagline(getString(c, COLUMN_TAGLINE, null));
        translation.setManifestFileName(getString(c, COLUMN_MANIFEST, null));
        translation.setPublished(getBool(c, COLUMN_PUBLISHED, DEFAULT_PUBLISHED));
        translation.setDownloaded(getBool(c, COLUMN_DOWNLOADED, false));
        translation.setLastAccessed(getDate(c, COLUMN_LAST_ACCESSED, DEFAULT_LAST_ACCESSED));

        return translation;
    }
}

File: library/db/src/main/kotlin/org/keynote/godtools/android/db/TranslationMapper.kt
package org.keynote.godtools.android.db

import android.content.ContentValues
import android.database.Cursor
import org.ccci.gto.android.common.util.database.getInt
import org.ccci.gto.android.common.util.database.getLocale
import org.ccci.gto.android.common.util.database.getString
import org.cru.godtools.model.Language
import org.cru.godtools.model.Translation
import org.keynote.godtools.android.db.Contract.TranslationTable.COLUMN_DESCRIPTION
import org.keynote.godtools.android.db.Contract.TranslationTable.COLUMN_DETAILS_BIBLE_REFERENCES
import org.keynote.godtools.android.db.Contract.TranslationTable.COLUMN_DETAILS_CONVERSATION_STARTERS
import org.keynote.godtools.android.db.Contract.TranslationTable.COLUMN_DETAILS_OUTLINE
import org.keynote.godtools.android.db.Contract.TranslationTable.COLUMN_DOWNLOADED
import org.keynote.godtools.android.db.Contract.TranslationTable.COLUMN_LANGUAGE
import org.keynote.godtools.android.db.Contract.TranslationTable.COLUMN_MANIFEST
import org.keynote.godtools.android.db.Contract.TranslationTable.COLUMN_NAME
import org.keynote.godtools.android.db.Contract.TranslationTable.COLUMN_PUBLISHED
import org.keynote.godtools.android.db.Contract.TranslationTable.COLUMN_TAGLINE
import org.keynote.godtools.android.db.Contract.TranslationTable.COLUMN_TOOL
import org.keynote.godtools.android.db.Contract.TranslationTable.COLUMN_VERSION

internal object TranslationMapper : BaseMapper<Translation>() {
    override fun mapField(values: ContentValues, field: String, obj: Translation) {
        when (field) {
            COLUMN_TOOL -> values.put(field, obj.toolCode)
            COLUMN_LANGUAGE -> values.put(field, serialize(obj.languageCode))
            COLUMN_VERSION -> values.put(field, obj.version)
            COLUMN_NAME -> values.put(field, obj.name)
            COLUMN_DESCRIPTION -> values.put(field, obj.description)
            COLUMN_TAGLINE -> values.put(field, obj.tagline)
            COLUMN_DETAILS_OUTLINE -> values.put(field, obj.toolDetailsOutline)
            COLUMN_DETAILS_BIBLE_REFERENCES -> values.put(field, obj.toolDetailsBibleReferences)
            COLUMN_DETAILS_CONVERSATION_STARTERS -> values.put(field, obj.toolDetailsConversationStarters)
            COLUMN_MANIFEST -> values.put(field, obj.manifestFileName)
            COLUMN_PUBLISHED -> values.put(field, true)
            COLUMN_DOWNLOADED -> values.put(field, obj.isDownloaded)
            else -> super.mapField(values, field, obj)
        }
    }

    override fun newObject(c: Cursor) = Translation()
    override fun toObject(c: Cursor) = super.toObject(c).apply {
        toolCode = c.getString(COLUMN_TOOL)
        languageCode = c.getLocale(COLUMN_LANGUAGE, Language.INVALID_CODE)
        version = c.getInt(COLUMN_VERSION, Translation.DEFAULT_VERSION)
        name = c.getString(COLUMN_NAME)
        description = c.getString(COLUMN_DESCRIPTION)
        tagline = c.getString(COLUMN_TAGLINE)
        toolDetailsOutline = c.getString(COLUMN_DETAILS_OUTLINE)
        toolDetailsBibleReferences = c.getString(COLUMN_DETAILS_BIBLE_REFERENCES)
        toolDetailsConversationStarters = c.getString(COLUMN_DETAILS_CONVERSATION_STARTERS)
        manifestFileName = c.getString(COLUMN_MANIFEST)
        isDownloaded = getBool(c, COLUMN_DOWNLOADED, false)
    }
}
